Implement JsGetProxyProperties API fixes #950

rhuanjl opened the pull request:
Responding to issue #950 

This PR:
1. Adds a JavascriptProxy::IsRevoked method to the Javascript proxy class (necessary for the below)
2. Adds a JsGetProxyProperties API to Jsrt which can:
    a) check if an object is a proxy -> set a provided bool to true/false
    b) if it is a proxy check if it's revoked
    c) if it is a revoked proxy set provided target and handler references to nullptr
    d) if it is a proxy that is not revoked provide references to it's target and handler
3. Tracks the same API through to WScriptJsrt in ch
4. Adds a test that uses the ch implementation

(Targeting 1.9 as this will assist with an issue in node-chakracore https://github.com/nodejs/node-chakracore/issues/488 )

**CC:** @jackhorton @kfarnung @dilijev

jackhorton commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1371,6 +1372,48 @@ JsValueRef __stdcall WScriptJsrt::SleepCallback(JsValueRef callee, bool isConstr
+                JsValueRef undefined = JS_INVALID_REFERENCE;
+                IfJsrtErrorSetGo(ChakraRTInterface::JsGetUndefinedValue(&undefined));
+                IfJsrtErrorSetGo(ChakraRTInterface::JsSetProperty(returnValue, targetProperty, undefined, true));
+                IfJsrtErrorSetGo(ChakraRTInterface::JsSetProperty(returnValue, handlerProperty, undefined, true));
```
Not sure if there is a benefit to doing this here -- the value of return.target will be undefined regardless of whether or not you set it. I personally would say that this method should return `{ target, handler}` or `undefined` rather than `{target, handler}` or `{}`

rhuanjl replied:
Note changed - I think there is potential utility to this - also it enables me to explicitly test the handling of a revoked proxy.

Currently this API does the following:
1. if given a non-proxy -> returns undefined
2. if given a revoked proxy -> returns { target : undefined, handler : undefined} -> you can explicitly detect a revoked proxy this way as the return value will have two properties not 0
3. if given a proxy that has not been revoked returns {target : proxyTarget, handler : proxyHandler}

jackhorton replied:
Sure, but then if nothing else, you still dont need to assign undefined to ret.target and ret.handler, they'll be undefined regardless.

fatcerberus replied:
@jackhorton They will, but then `Object.keys(ret).length` would then be 0 instead of 2 which seems to be what @rhuanjl was getting at.

rhuanjl replied:
I guess it could have {} as the return value for a revoked proxy - just seemed easier to explicitly test it this way.

boingoing replied:
The object returned from this helper can have a `revoked` property set to true if handler and target are null. We could use that to detect a revoked proxy. Would be easier to test that. Pretty sure real proxy can never have undefined or null handler and target unless it's revoked, though. I'm not really invested in either solution.

jackhorton commented on lib/Jsrt/Jsrt.cpp:
```diff
@@ -5329,4 +5329,39 @@ CHAKRA_API JsSetHostPromiseRejectionTracker(_In_ JsHostPromiseRejectionTrackerCa
+        VALIDATE_INCOMING_REFERENCE(object, scriptContext);
+        PARAM_NOT_NULL(isProxy);
+
+        *isProxy = false;
```
This initial assignment is probably unnecessary.

rhuanjl replied:
I was following the pseudo-policy I've seen of always initialising out values. But in this case considering it's set to a bool on the next line I can see your point - changed.

fatcerberus replied:
@rhuanjl The idea I think is that out parameters are always written to regardless of which codepath you take (in languages like C# with native support for out params it’s actually an error not to do so).  If, like here, the value is unconditionally set to something then initializing it is redundant.

jackhorton replied:
I think SAL/Prefast is supposed to give errors for things like that, but I am not sure if it catches everything.

jackhorton commented on lib/Jsrt/Jsrt.cpp:
```diff
@@ -5329,4 +5329,39 @@ CHAKRA_API JsSetHostPromiseRejectionTracker(_In_ JsHostPromiseRejectionTrackerCa
+
+        Js::JavascriptProxy* proxy = Js::JavascriptProxy::UnsafeFromVar(object);
+
+        if (target != nullptr && handler != nullptr)
```
Maybe check target and handler separately in case someone wanted one but not the other?

rhuanjl replied:
Good idea - done

jackhorton commented on lib/Runtime/Library/JavascriptProxy.h:
```diff
@@ -52,6 +52,7 @@ namespace Js
         JavascriptProxy(DynamicType * type, ScriptContext * scriptContext, RecyclableObject* target, RecyclableObject* handler);
         static BOOL Is(_In_ Var obj);
         static BOOL Is(_In_ RecyclableObject* obj);
+        BOOL IsRevoked();
```
nit: I personally think all new APIs should be `bool` rather than `BOOL`

rhuanjl replied:
I was just following what was done in the file, changed now though.

fatcerberus replied:
I don’t understand the motivation for returning `const bool`?  The `const` in this context seems redundant since `bool` is an immutable value type.  What am I missing?

jackhorton replied:
@fatcerberus I meant more like `bool IsRevoked() const;`, since it can be called on const proxies with no side effects.

rhuanjl replied:
oops fixed (and squashed)

jackhorton replied:
Also for clarification on the BOOL vs bool, BOOL was the old (microsoft-specific?) version before `bool` was part of the C++ standard (and I think its still not even in C, ridiculously). Its main problem is that its typedef'd to `int`, which causes silly conversion warnings among other things. Old APIs will remain BOOL indefinitely since changing it would affect almost every file (not to mention some uses of BOOL are for interaction with Win32 APIs that expect it), but I think we have been generally trying to avoid it for anything new unless it needs to interop.

fatcerberus replied:
C does have bool now, but you have to include stdbool.h for the typedef as it’s normally called `_Bool`.  I use it in my own code, it was a long overdue addition.

boingoing commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1371,6 +1372,48 @@ JsValueRef __stdcall WScriptJsrt::SleepCallback(JsValueRef callee, bool isConstr
+        bool isProxy = false;
+        JsValueRef target;
+        JsValueRef handler;
+        ChakraRTInterface::JsGetProxyProperties(arguments[1], &isProxy, &target, &handler);
```
Wrap in `IfJsrtErrorSetGo`?

boingoing commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1371,6 +1372,48 @@ JsValueRef __stdcall WScriptJsrt::SleepCallback(JsValueRef callee, bool isConstr
+            if (target == nullptr)
+            {
+                JsValueRef undefined = JS_INVALID_REFERENCE;
+                IfJsrtErrorSetGo(ChakraRTInterface::JsGetUndefinedValue(&undefined));
```
We already called `JsGetUndefinedValue` above. Can we fold these two into one call at the top?

boingoing commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1371,6 +1372,48 @@ JsValueRef __stdcall WScriptJsrt::SleepCallback(JsValueRef callee, bool isConstr
+            IfJsrtErrorSetGo(CreatePropertyIdFromString("handler", &handlerProperty));
+            IfJsrtErrorSetGo(ChakraRTInterface::JsCreateObject(&returnValue));
+
+            if (target == nullptr)
```
Check to see if `target == JS_INVALID_REFERENCE`.

boingoing commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1371,6 +1372,48 @@ JsValueRef __stdcall WScriptJsrt::SleepCallback(JsValueRef callee, bool isConstr
+            else
+            {
+                IfJsrtErrorSetGo(ChakraRTInterface::JsSetProperty(returnValue, targetProperty, target, true));
+                IfJsrtErrorSetGo(ChakraRTInterface::JsSetProperty(returnValue, handlerProperty, handler, true));
```
It's a little bit weird to me to set both of these properties in both sides of the conditional. What about setting `target` and `handler` to `undefined` if they're `JS_INVALID_REFERENCE` and having one set of `JsSetProperty` which always gets called?

boingoing commented on lib/Jsrt/ChakraCore.h:
```diff
@@ -1035,5 +1035,33 @@ CHAKRA_API
+///     If object is not a Proxy object the target and handler parameters are not touched.
+///     If nullptr is supplied for target or handler the function returns after
+///     setting the isProxy value.
+///     If the object is a revoked Proxy target and handler are made to point to nullptr.
```
I think we should set the target and handler to `JS_INVALID_REFERENCE` instead of nullptr and have this reflected in the doc here.

boingoing commented on lib/Jsrt/Jsrt.cpp:
```diff
@@ -5329,4 +5329,45 @@ CHAKRA_API JsSetHostPromiseRejectionTracker(_In_ JsHostPromiseRejectionTrackerCa
+
+        if (!*isProxy)
+        {
+            return JsNoError;
```
Unless I'm mistaken, `_Out_opt_` has the same semantics as `_Out_` when it comes to initialization. That means this function should not return without initializing `target` and `handler` if they are not null. We should just initialize both of them to `JS_INVALID_REFERENCE` before the assignment to `isProxy` if they're non-null.

fatcerberus replied:
Yeah, `_Out_opt` from what I've been able to gather just means the out pointer itself can be nullptr - if it's a non-null pointer then it should indeed be treated the same as an `_Out_` parameter.

boingoing commented on test/es6/ProxyPropertiesAPI.js:
```diff
@@ -0,0 +1,85 @@
+            assert.isTrue(names.includes("handler"));
+
+            assert.isUndefined(properties.target, "Target of revoked proxy should be undefined.");
+            assert.isUndefined(properties.handler, "Handler of revoekd proxy should be undefined.");
```
Typo in revoked.

rhuanjl commented:
@boingoing Thanks for the review I think I've addressed all of your points. (Done in a separate commit now to make it easy to compare)

boingoing commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1371,6 +1372,57 @@ JsValueRef __stdcall WScriptJsrt::SleepCallback(JsValueRef callee, bool isConstr
+
+            if (target == JS_INVALID_REFERENCE)
+            {
+                ChakraRTInterface::JsGetTrueValue(&revoked);
```
For consistency, can we wrap this in `IfJsrtErrorSetGo`?

boingoing commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1371,6 +1372,57 @@ JsValueRef __stdcall WScriptJsrt::SleepCallback(JsValueRef callee, bool isConstr
+            }
+            else
+            {
+                ChakraRTInterface::JsGetFalseValue(&revoked);
```
Same nitpick about `IfJsrtErrorSetGo`.

kfarnung commented on lib/Runtime/Library/JavascriptProxy.h:
```diff
@@ -52,6 +52,7 @@ namespace Js
         JavascriptProxy(DynamicType * type, ScriptContext * scriptContext, RecyclableObject* target, RecyclableObject* handler);
         static BOOL Is(_In_ Var obj);
         static BOOL Is(_In_ RecyclableObject* obj);
+        bool IsRevoked() const;
```
nit: I'd move this down with the other non-static, non-virtual methods.

rhuanjl commented:
@boingoing Thanks for the review and sorry for the mistakes. I've fixed those additional points, updated to latest and squashed. Please let me know if anything else is needed.

rhuanjl commented:
@dotnet-bot test Windows ci_slow_x64_debug please

akroshg commented on test/es6/ProxyPropertiesAPI.js:
```diff
@@ -0,0 +1,90 @@
+            let properties = WScript.GetProxyProperties(proxy);
+
+            let names = Object.getOwnPropertyNames(properties);
+            assert.areEqual(names.length, 3, "proxy properties names should have length 2");
```
>length 2 [](start = 81, length = 8)

shoul have 3 ?

akroshg replied:
should*

---
In reply to: [174286438](https://github.com/Microsoft/ChakraCore/pull/4806#discussion_r174286438) [](ancestors = 174286438)

boingoing commented:
@rhuanjl Thanks once again for your contribution here! I'll start the internal testing process for this one.

rhuanjl commented:
@boingoing thanks for getting this merged I hope it's helpful.
